pinot-broker/src/main/java/org/apache/pinot/broker/routing/v2/RoutingManager.java

```diff
@@ -0,0 +1,481 @@
+
+    Map<String, String> segmentToInstanceMap = routingEntry.calculateSegmentToInstanceMap(brokerRequest);
+    Map<ServerInstance, List<String>> routingTable = new HashMap<>();
+    for (Map.Entry<String, String> entry : segmentToInstanceMap.entrySet()) {
+      ServerInstance serverInstance = _enabledServerInstanceMap.get(entry.getValue());
+      if (serverInstance != null) {
+        routingTable.computeIfAbsent(serverInstance, k -> new ArrayList<>()).add(entry.getKey());
+      } else {
+        // Should not happen in normal case unless encountered unexpected exception when updating routing entries
+        _brokerMetrics.addMeteredTableValue(tableNameWithType, BrokerMeter.SERVER_MISSING_FOR_ROUTING, 1L);
+      }
+    }
+    return routingTable;
+  }
+
+  /**
+   * Returns the time boundary info for the given offline table name.
+   * <p>NOTE: Time boundary info is only available for the offline part of the hybrid table.
+   */
+  @Nullable
+  public TimeBoundaryInfo getTimeBoundaryInfo(String offlineTableName) {
+    RoutingEntry routingEntry = _routingEntryMap.get(offlineTableName);
+    Preconditions.checkState(routingEntry != null, "Failed to find routing for table: %s", offlineTableName);
+    TimeBoundaryManager timeBoundaryManager = routingEntry.getTimeBoundaryManager();
+    return timeBoundaryManager != null ? timeBoundaryManager.getTimeBoundaryInfo() : null;
+  }
+
+  private static class RoutingEntry {
+    final String _tableNameWithType;
+    final SegmentSelector _segmentSelector;
+    final SegmentPruner _segmentPruner;
+    final InstanceSelector _instanceSelector;
+
+    // Cache the ExternalView version for the last update
+    transient int _lastUpdateExternalViewVersion;
+    // Time boundary manager is only available for the offline part of the hybrid table
+    transient TimeBoundaryManager _timeBoundaryManager;
+
+    RoutingEntry(String tableNameWithType, SegmentSelector segmentSelector, SegmentPruner segmentPruner,
+        InstanceSelector instanceSelector, int lastUpdateExternalViewVersion,
+        @Nullable TimeBoundaryManager timeBoundaryManager) {
+      _tableNameWithType = tableNameWithType;
+      _segmentSelector = segmentSelector;
+      _segmentPruner = segmentPruner;
+      _instanceSelector = instanceSelector;
+      _lastUpdateExternalViewVersion = lastUpdateExternalViewVersion;
+      _timeBoundaryManager = timeBoundaryManager;
+    }
+
+    String getTableNameWithType() {
+      return _tableNameWithType;
+    }
+
+    int getLastUpdateExternalViewVersion() {
+      return _lastUpdateExternalViewVersion;
+    }
+
+    void setTimeBoundaryManager(@Nullable TimeBoundaryManager timeBoundaryManager) {
+      _timeBoundaryManager = timeBoundaryManager;
+    }
+
+    @Nullable
+    TimeBoundaryManager getTimeBoundaryManager() {
+      return _timeBoundaryManager;
+    }
+
+    void onExternalViewChange(ExternalView externalView, Set<String> onlineSegments) {
```

Comment:
What if `calculateSegmentToInstanceMap()` got hit while we are processing `onExternalViewChange()` in the middle? (e.g. segment selector/pruner got updated while instance selector not updated). Do we get impact on the final result? - I'm pretty sure you already thought about this :)

Reply:
Yes, we won't route to the new segments before the whole routing entry gets updated. Will add some comments here